src/Libraries/DSCPython/DCPythonCodeCompletionProviderCore.cs

```diff
@@ -0,0 +1,806 @@
+                using (Py.GIL())
+                {
+                    if (code.Contains("\"\"\""))
+                    {
+                        code = PythonCodeCompletionUtils.StripDocStrings(code);
+                    }
+
+                    if (Scope == null)
+                    {
+                        Scope = CreateUniquePythonScope();
+                    }
+
+                    UpdateImportedTypes(code);
+                    UpdateVariableTypes(code);  // Possibly use hindley-milner in the future
+
+                    // If expand param is true use the entire namespace from the line of code
+                    // Else just return the last name of the namespace
+                    string name = expand ? PythonCodeCompletionUtils.GetLastNameSpace(code) : PythonCodeCompletionUtils.GetLastName(code);
+                    if (!String.IsNullOrEmpty(name))
+                    {
+                        try
+                        {
+                            // Attempt to get type using naming
+                            Type type = expand ? PythonCodeCompletionUtils.TryGetTypeFromFullName(name) : TryGetType(name);
+
+                            // CLR type
+                            if (type != null)
+                            {
+                                items = EnumerateMembers(type, name).Select(x => new PythonCodeCompletionDataCore(x.Item1, x.Item2, x.Item3, x.Item4, this));
+                            }
+                            // Variable type
+                            else if (VariableTypes.TryGetValue(name, out type))
+                            {
+                                items = EnumerateMembers(type, name).Select(x => new PythonCodeCompletionDataCore(x.Item1, x.Item2, x.Item3, x.Item4, this));
+                            }
+                            else
+                            {
+                                var mem = LookupMember(name);
+                                var pythonModule = mem as PyObject;
+                                // Python Module type
+                                if (pythonModule != null)
+                                {
+                                    items = EnumerateMembers(pythonModule, name).Select(x => new PythonCodeCompletionDataCore(x.Item1, x.Item2, x.Item3, x.Item4, this));
+                                }
+                            }
+                        }
+                        catch (Exception ex)
+                        {
+                            Log(ex.ToString());
+                        }
+                    }
+                }
+            }
+            finally
+            {
+                PythonEngine.ReleaseLock(gs);
+            }
+            // If unable to find matching results and expand was set to false,
+            // try again using the full namespace (set expand to true)
+            if (!items.Any() && !expand)
+            {
+                return GetCompletionData(code, true);
```

Comment:
could this end up recursing infinitely?

Reply:
Only 1 more recursion is possible because of the `expand` flag